edgedetector/solver/evaluator/cost/curvature/curvature_cost_evaluator.py

```diff
@@ -15,9 +15,34 @@ def evaluate_window(self, row, column):
             return 1
         if len(neighbours) <= 1:
             return 0
-        [(x1, y1), (x2, y2)] = neighbours
-        if x1 != x2 and y1 != y2:
-            return 1
-        return 0.5
+        [n1, n2] = neighbours
+        if __class__.__turns_by_45((row, column), n1, n2):
+            return 0.5
+        if __class__.__inline((row, column), n1, n2):
+            return 0
+        return 1
+
+    @staticmethod
+    def __turns_by_45(center, p1, p2):
+        return (__class__.__turns_by_45_from_line(center, p1, p2) or
+                __class__.__turns_by_45_from_line(center, p2, p1))
 
+    @staticmethod
+    def __turns_by_45_from_line(center, p1, p2):
+        """Checks if p2 turns by 45 degrees from line
+        determined by center and p1"""
+        (x1, y1) = center
+        (x2, y2) = p1
+        (x3, y3) = p2
+        return ((x1 == x2 or y1 == x2) and
+                x3 != x1 and y3 != y1 and x3 != x2 and y3 != y2)
 
+    @staticmethod
+    def __inline(p1, p2, p3):
+        (x1, y1) = p1
+        (x2, y2) = p2
+        (x3, y3) = p3
+        return ((x1 == x2 == x3) or                         # horizontal
+                (y1 == y2 == y3) or                         # vertical
```

Comment:
first index corresponds to row, second to column ( as usual with 2D arrays)
x,y are not like in carthesian coordinate system